File: Day4/Day4.py
import pathlib


def parse(puzzle_input):
    values = puzzle_input.split("\n")
    data = []
    for value in values:
        first, second = value.split(",")
        one, two = first.split("-")
        three, four = second.split("-")
        data.append([(int(one), int(two)), (int(three), int(four))])
    return data


def part1(data):
    fullycontains = 0
    for assignment in data:
        ass = assignment[0]
        ign = assignment[1]
        fullycontains = fulloverlap(ass, ign)
    return fullycontains


def part2(data):
    contains = 0
    for assignment in data:
        ass = assignment[0]
        ign = assignment[1]
        contains += overlap(ass, ign)
    return contains


def solve(puzzle_input):
    data = parse(puzzle_input)
    sol1 = part1(data)
    sol2 = part2(data)
    return sol1, sol2


def fulloverlap(tuple1, tuple2):
    set1 = set(list(range(tuple1[0], tuple1[1]+1)))
    set2 = set(list(range(tuple2[0], tuple2[1]+1)))
    return set1.issubset(set2) or set2.issubset(set1)


def overlap(tuple1, tuple2):
    set1 = set(list(range(tuple1[0], tuple1[1]+1)))
    set2 = set(list(range(tuple2[0], tuple2[1]+1)))
    overlapset = set1.intersection(set2)
    return not len(overlapset) == 0


def run():
    puzzle_input = pathlib.Path("input").read_text().strip()
    solutions = solve(puzzle_input)
    print(solutions)


run()

File: 2022/Day4/Day4.R
input <- readLines("input")
fullycontained <- 0
partlycontained <- 0
for (assignments in input) {
  asses <- strsplit(assignments, ",")[[1]]
  ass <- strsplit(asses[1], "-")[[1]]
  one <- as.numeric(ass[1])
  two <- as.numeric(ass[2])
  ign <- strsplit(asses[2], "-")[[1]]
  three <- as.numeric(ign[1])
  four <- as.numeric(ign[2])
  if ((one >= three && four >= two) || (one <= three && four <= two)) {
    fullycontained <- fullycontained + 1
  }
  if ((max(one, three) <= min(two, four))) {
    partlycontained <- partlycontained + 1
  } 
}
print(fullycontained)
print(partlycontained)
